contents/dataframes_performance.md

```diff
@@ -4,8 +4,240 @@ So far, we didn't think about making our `DataFrames.jl` code **fast**.
+
+```jl
+s = """
+    # non allocarting function # hide
+    select!(responses(), :id, :q1)
+    """
+sco(s; process=without_caption_label)
+```
+
+The `@allocated` macro tells us how much memory was allocated.
+In other words, how much new information the computer had to store in its memory while running the code.
+Let's see how they will perform: 
+
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated select(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated select!(df, :id, :q1)
+    """
+sco(s; process=string, post=plainblock)
+```
+
+As we can see, `select!` allocates less than `select`.
+So, it should be faster and while consuming less memory.
+
+### Copying vs Not Copying Columns
+
+There are **two ways to access a DataFrame column**.
+They differ in how they are accessed: one creates a "view" to the column without copying and the other creates a whole new column by copying the original column.
+
+The first is the regular dot `.` operator followed by the column name, like in `df.col`.
+This kind of access **does not copy** the column `col`.
+Instead `df.col` creates a "view" which is a link to the original column without performing any allocation.
+Additionally, the syntax `df.col` is the same as `df[!, :col]` with the bang `!` as the row selector.
+
+The second way to access a `DataFrame` column is the `df[:, :col]` with the colon `:` as the row selector.
+This kind of access **does copy** the column `col`, so beware that it may produce unwanted allocations.
+
+As before, let's try out our two ways to access a column in the responses `DataFrame`:
+
+```jl
+s = """
+    # allocation # hide
+    df = responses()
+    @allocated col = df[:, :id]
+    """
+sco(s; process=string, post=plainblock)
 ```
+
+
+```jl
+s = """
+    # non allocation # hide
+    df = responses()
+    @allocated col = df[!, :id]
+    """
+sco(s; process=string, post=plainblock)
+```
+
+When we access a column without copying it we are making zero allocations and our code should be faster.
+So, if you don't need a copy always access your `DataFrame`s columns with `df.col` or `df[!, :col]` instead of `df[:, :col]`.
+
+### CSV.read versus CSV.File
+
+If you take a look at the help output for `CSV.read`, you will see that it is a convenience function identical to the function called `CSV.File` with the same keywords arguments.
+Both `CSV.read` and `CSV.File` will read the contents of a CSV file, but they differ in the default behavior.
+**`CSV.read`, by default, will not make copies** of the incoming data.
+Instead, `CSV.read` will pass all the data to the second argument (known as the "sink").
+
+So, something like this:
+
+```julia
+df = CSV.read("file.csv", DataFrame)
+```
+
+will pass all the incoming data from `file.csv` to the `DataFrame` sink, thus returning a `DataFrame` type that we store in the `df` variable.
```

Comment:
is this 100% true? when doing the passing is not doing the copying to the DataFrame as well?

Reply:
Yes taken strictly from the CSV.jl documentation and some discourse threads